File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    static List<Integer>[] graph;
    static boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        init(n, computers);
        
        for (int i = 1; i < graph.length; i++) {
            if (!visited[i]) {
                answer++;
                dfs(i);
            }
        }
        
        return answer;
    }
    
    private void init(int n, int[][] computers) {
        graph = new ArrayList[n + 1];
        visited = new boolean[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < computers.length; i++) {
            List<Integer> linkedComputer = graph[i + 1];
            for (int j = 0; j < computers[0].length; j++) {
                if (i == j) {
                    continue;
                }
                if (computers[i][j] == 1) {
                    linkedComputer.add(j + 1);
                }
            }
        }
    }
    
    private void dfs(int computer) {
        visited[computer] = true;
        List<Integer> linkedComputer = graph[computer];
        for (int next : linkedComputer) {
            if (!visited[next]) {
                dfs(next);
            } 
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0
    
    graph = {}
    visited = [False] * n
    for i in range(len(computers)):
        graph[i] = []
        computer = computers[i]
        for j in range(len(computer)):
            if i == j:
                continue
            connect = computer[j]
            if connect == 1:
                graph[i].append(j)
                
    for i in range(0, n):
        if not visited[i]:
            answer += 1
            visited[i] = True
            q = deque()
            q.append(i)
    
            while q:
                p = q.popleft()
                
                nexts = graph[p]
                
                for next in nexts:
                    if not visited[next]:
                        visited[next] = True
                        q.append(next)
    
    return answer